README.md
# Tokio Compat

Compatibility layers between `tokio` 0.2 and legacy versions.

[![Crates.io][crates-badge]][crates-url]
[![Documentation][docs-badge]][docs-url]
[![MIT licensed][mit-badge]][mit-url]
[![Build Status][actions-badge]][actions-url]
[![Discord][discord-badge]][discord-url]

[Website](https://tokio.rs) |
[Guides](https://tokio.rs/docs/) |
[API Docs][docs-url] |
[Chat][discord-url]

[crates-badge]: https://img.shields.io/crates/v/tokio-compat.svg
[crates-url]: https://crates.io/crates/tokio-compat
[docs-url]: https://docs.rs/tokio-compat/0.1.6/tokio_compat
[docs-badge]: https://docs.rs/tokio-compat/badge.svg
[mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg
[mit-url]: LICENSE
[actions-badge]: https://github.com/tokio-rs/tokio-compat/workflows/CI/badge.svg
[actions-url]: https://github.com/tokio-rs/tokio-compat/actions?query=workflow%3ACI
[discord-badge]: https://img.shields.io/discord/500028886025895936?logo=discord&label=discord&logoColor=white
[discord-url]: https://discord.gg/6yGkFeN

## Overview

This crate provides compatibility runtimes that allow running both `futures` 0.1
futures that use `tokio` 0.1 runtime services _and_ `std::future` futures that
use `tokio` 0.2 runtime services.

### Examples

Spawning both `tokio` 0.1 and `tokio` 0.2 futures:

```rust
use futures_01::future::lazy;

tokio_compat::run(lazy(|| {
    // spawn a `futures` 0.1 future using the `spawn` function from the
    // `tokio` 0.1 crate:
    tokio_01::spawn(lazy(|| {
        println!("hello from tokio 0.1!");
        Ok(())
    }));

    // spawn an `async` block future on the same runtime using `tokio`
    // 0.2's `spawn`:
    tokio_02::spawn(async {
        println!("hello from tokio 0.2!");
    });

    Ok(())
}))
```

Futures on the compat runtime can use `timer` APIs from both 0.1 and 0.2
versions of `tokio`:

```rust
use std::time::{Duration, Instant};
use futures_01::future::lazy;
use tokio_compat::prelude::*;

tokio_compat::run_std(async {
    // Wait for a `tokio` 0.1 `Delay`...
    let when = Instant::now() + Duration::from_millis(10);
    tokio_01::timer::Delay::new(when)
        // convert the delay future into a `std::future` that we can `await`.
        .compat()
        .await
        .expect("tokio 0.1 timer should work!");
    println!("10 ms have elapsed");

    // Wait for a `tokio` 0.2 `Delay`...
    let when = Instant::now() + Duration::from_millis(20);
    tokio_02::timer::delay(when).await;
    println!("20 ms have elapsed");
});
```

## License

This project is licensed under the [MIT license](LICENSE).

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in Tokio by you, shall be licensed as MIT, without any additional
terms or conditions.


CHANGELOG.md
# 0.1.6 (April 21, 2020)

### Changed

- Make documentation for the `prelude` module properly link to reexports (#31)

# 0.1.5 (March 3, 2020)

### Fixed

- FD leak caused by leaky strong Arc reference (#28)

# 0.1.4 (January 8, 2020)

### Fixed

- Compiler warnings in the threadpool runtime (#21)

# 0.1.3 (January 7, 2020)

### Fixed

- Removed unnecessary `'static` bounds on the `current_thread` runtime's
  `block_on`/`block_on_std` functions that broke API compatibility with Tokio
  0.1 (#17)

### Added

- Implementation of `futures` 0.1 `Executor` trait for `TaskExecutor`s (#18)

# 0.1.2 (December 18, 2019)

### Fixed

- Dependencies on `futures-core-preview` and `futures-util-preview` updated to
  release `futures` 0.3 versions (#14)

# 0.1.1 (December 18, 2019)

### Fixed

- Hang when calling `block_on` twice on the same `Runtime` (#11)
- `Runtime::shutdown_on_idle` and `current_thread::Runtime::run` completing
  early when a future has completed previously on that runtime (#12)

# 0.1.0 (December 17, 2019)

- Initial release


src/lib.rs
//! Compatibility between `tokio` 0.2 and legacy versions.
//!
//! ## Overview
//!
//! This crate provides compatibility runtimes that allow running both `futures` 0.1
//! futures that use `tokio` 0.1 runtime services _and_ `std::future` futures that
//! use `tokio` 0.2 runtime services.
//!
//! ### Examples
//!
//! Spawning both `tokio` 0.1 and `tokio` 0.2 futures:
//!
//! ```rust
//! use futures_01::future::lazy;
//!
//! tokio_compat::run(lazy(|| {
//!     // spawn a `futures` 0.1 future using the `spawn` function from the
//!     // `tokio` 0.1 crate:
//!     tokio_01::spawn(lazy(|| {
//!         println!("hello from tokio 0.1!");
//!         Ok(())
//!     }));
//!
//!     // spawn an `async` block future on the same runtime using `tokio`
//!     // 0.2's `spawn`:
//!     tokio_02::spawn(async {
//!         println!("hello from tokio 0.2!");
//!     });
//!
//!     Ok(())
//! }))
//! ```
//!
//! Futures on the compat runtime can use `timer` APIs from both 0.1 and 0.2
//! versions of `tokio`:
//!
//! ```rust
//! use std::time::{Duration, Instant};
//! use tokio_compat::prelude::*;
//!
//! tokio_compat::run_std(async {
//!     // Wait for a `tokio` 0.1 `Delay`...
//!     let when = Instant::now() + Duration::from_millis(10);
//!     tokio_01::timer::Delay::new(when)
//!         // convert the delay future into a `std::future` that we can `await`.
//!         .compat()
//!         .await
//!         .expect("tokio 0.1 timer should work!");
//!     println!("10 ms have elapsed");
//!
//!     // Wait for a `tokio` 0.2 `Delay`...
//!     tokio_02::time::delay_for(Duration::from_millis(20)).await;
//!     println!("20 ms have elapsed");
//! });
//! ```
//!
//! ## Feature Flags
//!
//! - `rt-current-thread`: enables the `current_thread` compatibilty runtime
//! - `rt-full`: enables the `current_thread` and threadpool compatibility
//!   runtimes (enabled by default)
#![doc(html_root_url = "https://docs.rs/tokio-compat/0.1.6")]
#![warn(
    missing_debug_implementations,
    missing_docs,
    rust_2018_idioms,
    unreachable_pub
)]
#![deny(intra_doc_link_resolution_failure)]
#![doc(test(
    no_crate_inject,
    attr(deny(warnings, rust_2018_idioms), allow(dead_code, unused_variables))
))]
#![cfg_attr(docsrs, feature(doc_cfg))]

#[cfg(any(feature = "rt-current-thread", feature = "rt-full"))]
#[cfg_attr(
    docsrs,
    doc(cfg(any(feature = "rt-current-thread", feature = "rt-full")))
)]
pub mod runtime;

#[cfg(feature = "rt-full")]
pub use self::runtime::{run, run_std};
pub mod prelude;
